### PR TITLE
Keep reference to `DomainParticipantFactory`

### PR DESCRIPTION
**Pull request by MiguelCompany**
We've seen several issues related with destruction order lately.

This PR aims to mitigate them by leveraging Fast DDS `DomainParticipantFactory::get_shared_instance()`.

It should then avoid the automatic destruction of the factory (and the participants it created) until all the contexts in the process have been destroyed.

This can be backported to Jazzy and Iron, but not Humble, since `get_shared_instance()` is only available from Fast DDS 2.8.0 onwards.

### CODE REVIEW

File: rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/custom_participant_info.hpp
Comment (fujitatomoya) [inline rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/custom_participant_info.hpp:97]: > It should then avoid the automatic destruction of the factory (and the participants it created) until all the contexts in the process have been destroyed.

this will `eprosima::fastdds::dds::DomainParticipantFactory` object until all contexts in the process space are destroyed for sure. but i am not sure if this works for `eprosima::fastdds::dds::DomainParticipant`... does this mean `eprosima::fastdds::dds::DomainParticipantFactory` keeps the shared_ptr to `eprosima::fastdds::dds::DomainParticipant` until dtor?

Comment (fujitatomoya) [reply rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/custom_participant_info.hpp:97]: @MiguelCompany thanks for the info.

although i understand the code stack, i am not sure how / what this is trying to mitigate?

> This PR aims to mitigate them

can you elaborate this `them` issues in example if possible?

Comment (MiguelCompany) [reply rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/custom_participant_info.hpp:97]: In short: Any issue related with call stacks that include `__cxa_finalize()`

Recent examples:
* https://github.com/ros2/rmw_fastrtps/issues/761
* https://github.com/ros2/rmw_fastrtps/issues/760

Old (long standing) issues:
* https://github.com/ros2/rclcpp/issues/1766
* https://github.com/ros2/ros2/issues/1439

Comment (MiguelCompany) [reply rmw_fastrtps_shared_cpp/include/rmw_fastrtps_shared_cpp/custom_participant_info.hpp:97]: Additionally, this was discussed with @ivanpauno [here](https://github.com/eProsima/Fast-DDS/discussions/2537) when the context concept was introduced

**Review by fujitatomoya — APPROVED** [review]
@MiguelCompany thanks for the explanation, lgtm with green CI.

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21296/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/15615/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/796/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22030/)

**Comment by MiguelCompany** [comment]
@fujitatomoya @ahcorde I rebased this one to get the changes from #763 (I think the test failures were due to that)

**Comment by ahcorde** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21982/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16247/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1496/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22712/)

**Review by clalancette — APPROVED** [review]

**Comment by clalancette** [comment]
I'm going to run one more CI here, just to make sure everything is good, before merging.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21984/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16249/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1498/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22714/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/22715/)

**Comment by clalancette** [comment]
Ah, CI isn't working because this needs to be rebased.  @MiguelCompany can you do that?

**Comment by MiguelCompany** [comment]
> Ah, CI isn't working because this needs to be rebased. @MiguelCompany can you do that?

Done!

**Comment by clalancette** [comment]
Here we go:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/21991/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/16254/)
* Linux-rhel [Build Status](http://ci.ros2.org/job/ci_linux-rhel/1503/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/22724/)
* Windows Debug [Build Status](http://ci.ros2.org/job/ci_windows/22725/)

**Comment by clalancette** [comment]
The failures here are all in known flakes, so going ahead and merging this.

**Comment by MiguelCompany** [comment]
@clalancette @fujitatomoya Should we backport this?

**Comment by fujitatomoya** [comment]
@MiguelCompany i say we should especially for jazzy and humble, but this is breaking ABI change?

**Comment by MiguelCompany** [comment]
> I say we should especially for jazzy and humble, but this is breaking ABI change?

@fujitatomoya Strictly speaking, this is an ABI break. But since it is on the internal package `rmw_fastrtps_shared_cpp`, which is only used by the other two packages, it would not be a problem if we release new versions of the three packages in this repo at the same time.

**Comment by MiguelCompany** [comment]
BTW, it cannot be backported to Humble, because `get_shared_instance()` was added in Fast DDS 2.8.0, and Humble uses 2.6.x

**Comment by fujitatomoya** [comment]
@MiguelCompany i am okay with backport for jazzy, lets hear from other opinion.

**Comment by gavanderhoorn** [comment]
> @MiguelCompany i am okay with backport for jazzy, lets hear from other opinion.

If this hasn't yet been backported, I would really appreciate a backport to Jazzy.

**Comment by fujitatomoya** [comment]
@clalancette what do you think about backporting jazzy? IMO, this should be no problem and better bug fix for jazzy LTS.

**Comment by clalancette** [comment]
> @clalancette what do you think about backporting jazzy? IMO, this should be no problem and better bug fix for jazzy LTS.

I'm not entirely sure, but I think this may be involved in what is causing our crashes on RHEL with flake8.  I'm still debugging that, but commenting out the call to `delete_participant` here makes the crash stop happening.

Thus I think we should hold off on backporting this until we figure out what that problem is.

**Pull request merged**
